server/src/handlers/http/external.rs

```diff
@@ -0,0 +1,235 @@
+) -> Result<impl Responder, ModuleError> {
+    let list: Vec<String> = registry
+        .read()
+        .unwrap()
+        .registrations()
+        .map(|x| x.id.clone())
+        .collect();
+
+    Ok(web::Json(list))
+}
+
+pub async fn get_config(path: web::Path<(String, String)>) -> Result<impl Responder, ModuleError> {
+    let (name, stream_name) = path.into_inner();
+    let mut metadata = CONFIG
+        .storage()
+        .get_object_store()
+        .get_stream_metadata(&stream_name)
+        .await?;
+
+    let config = metadata
+        .modules
+        .remove(&name)
+        .ok_or_else(|| ModuleError::ModuleNotFound(name.clone()))?;
+
+    Ok(web::Json(config))
+}
+
+pub async fn put_config(
+    path: web::Path<(String, String)>,
+    config: web::Json<serde_json::Value>,
+    registry: web::Data<RwLock<ModuleRegistry>>,
+) -> Result<impl Responder, ModuleError> {
+    let (name, stream_name) = path.into_inner();
+    let config = config.into_inner();
+
+    let registration = registry
+        .read()
+        .unwrap()
+        .registrations()
+        .find(|&registration| registration.id == name)
+        .cloned()
+        .ok_or_else(|| ModuleError::ModuleNotFound(name.clone()))?;
+
+    let path = registration
+        .stream_config
+        .path
+        .replace("{stream_name}", &stream_name);
+
+    let url = registration.url.join(&path)?;
+
+    let client = reqwest::Client::new();
+    let req_body = serde_json::to_vec(&config).expect("valid json");
+
+    let request = client
+        .put(url)
+        .body(req_body)
+        .basic_auth(registration.username, Some(registration.password))
+        .header(reqwest::header::CONTENT_TYPE, "application/json") // Set the Content-Type header
+        .build()
+        .unwrap();
+    let resp = client.execute(request).await?;
+
+    if !resp.status().is_success() {
+        let resp_body = String::from_utf8_lossy(&resp.bytes().await?).to_string();
+        return Err(ModuleError::Custom(resp_body.into()));
+    }
+
+    CONFIG
+        .storage()
+        .get_object_store()
+        .put_module_config(&stream_name, name, config)
+        .await?;
+
+    Ok(HttpResponse::Ok())
+}
+
+pub async fn router(
+    req: HttpRequest,
+    params: web::Path<(String, String)>,
+    body: web::Bytes,
+    registry: web::Data<RwLock<ModuleRegistry>>,
+) -> Result<HttpResponse, ModuleError> {
+    let (name, path) = params.into_inner();
+    let method: &http::Method = req.method();
+
+    let (url, username, password) = {
+        let module_registry = registry.read().unwrap();
+        let registration = module_registry
+            .get(&name)
+            .ok_or_else(|| ModuleError::ModuleNotFound(name.clone()))?;
+
+        let module_path = registration.get_module_path(&path, method);
+        let module_path = module_path.unwrap_or_else(|| "".to_string());
+
+        if module_path.is_empty() {
+            return Ok(HttpResponse::NotFound().finish());
```

Comment:
@trueleo please see if this approach is fine.

Reply:
Refactored this part.  please take a look

Reply:
This seems much cleaner!💯